packages/smooth_app/lib/data_models/product_query_model.dart

```diff
@@ -41,17 +43,58 @@ class ProductQueryModel with ChangeNotifier {
   String? get loadingError => _loadingError;
   LoadingStatus get loadingStatus => _loadingStatus;
 
+  void _clear() {
+    currentCategory = _CATEGORY_ALL;
+    _loadingStatus = LoadingStatus.LOADING;
+    _loadingError = null;
+    _products.clear();
+    displayProducts = null;
+    categories.clear();
+    _categoriesCounter.clear();
+    sortedCategories.clear();
+  }
+
   Future<void> _asyncLoad() async {
     _loadingError = await supplier.asyncLoad();
     if (_loadingError != null) {
       _loadingStatus = LoadingStatus.ERROR;
     } else {
       _loadingStatus = LoadingStatus.LOADED;
-      _products = supplier.getProductList().getList();
+      _products.addAll(supplier.partialProductList.getProducts());
     }
     notifyListeners();
   }
 
+  Future<bool> loadNextPage() async {
+    final ProductListSupplier? refreshSupplier = supplier.getRefreshSupplier();
+    if (refreshSupplier != null) {
+      // in that case, we were on a database supplier, on an empty page
+      _supplier = refreshSupplier;
+    } else {
+      // in that case, we were on a back-end supplier, on a loaded page
+      supplier.productQuery.toNextPage();
+    }
+    await _asyncLoad();
+    return true;
+  }
+
+  // TODO(monsieurtanuki): don't clear everything if it fails?
+  Future<bool> loadFromTop() async {
+    _clear();
```

Comment:
Yeah we should probably keep some things like the category here, but thats not so important for now